### ISSUE TITLE
Serializing and deserializing Objects with a type attribute

### ISSUE THREAD

**Issue by ironhouzi** [marshmallow-code/marshmallow#491]
Given the class:

``` python
class Foo(object):
    def __init__(self, datatype):
        self.datatype = datatype

```

and the Schema:

``` python
class FooSchema(Schema):
    Meta:
        fields = ('datatype',)
```

I do:

```
>>> foo = Foo(datetime.datetime)
>>> schema = FooSchema()
>>> data, errors = schema.dump(foo)
```

Error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/utils.py in _get_value_for_key(key, obj, default)
    318     try:
--> 319         return obj[key]
    320     except (KeyError, AttributeError, IndexError, TypeError):

TypeError: 'Foo' object is not subscriptable

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-62-5580971ff97a> in <module>()
----> 1 data, errors = schema.dump(baz)

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/schema.py in dump(self, obj, many, update_fields, **kwargs)
    494                 dict_class=self.dict_class,
    495                 index_errors=self.opts.index_errors,
--> 496                 **kwargs
    497             )
    498         except ValidationError as error:

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/marshalling.py in serialize(self, obj, fields_dict, many, accessor, dict_class, index_errors, index)
    144                 field_name=key,
    145                 field_obj=field_obj,
--> 146                 index=(index if index_errors else None)
    147             )
    148             if value is missing:

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/marshalling.py in call_and_store(self, getter_func, data, field_name, field_obj, index)
     65         """
     66         try:
---> 67             value = getter_func(data)
     68         except ValidationError as err:  # Store validation errors
     69             self.error_kwargs.update(err.kwargs)

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/marshalling.py in <lambda>(d)
    138             key = ''.join([self.prefix or '', field_obj.dump_to or attr_name])
    139
--> 140             getter = lambda d: field_obj.serialize(attr_name, d, accessor=accessor)
    141             value = self.call_and_store(
    142                 getter_func=getter,

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/fields.py in serialize(self, attr, obj, accessor)
    239         """
    240         if self._CHECK_ATTRIBUTE:
--> 241             value = self.get_value(attr, obj, accessor=accessor)
    242             if value is missing_:
    243                 if hasattr(self, 'default'):

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/fields.py in get_value(self, attr, obj, accessor, default)
    183         accessor_func = accessor or utils.get_value
    184         check_key = attr if attribute is None else attribute
--> 185         return accessor_func(check_key, obj, default)
    186
    187     def _validate(self, value):

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/schema.py in get_attribute(self, attr, obj, default)
    402         .. versionadded:: 2.0.0
    403         """
--> 404         return utils.get_value(attr, obj, default)
    405
    406     ##### Handler decorators (deprecated) #####

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/utils.py in get_value(key, obj, default)
    304         return _get_value_for_key(key, obj, default)
    305     else:
--> 306         return _get_value_for_keys(key.split('.'), obj, default)
    307
    308

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/utils.py in _get_value_for_keys(keys, obj, default)
    309 def _get_value_for_keys(keys, obj, default):
    310     if len(keys) == 1:
--> 311         return _get_value_for_key(keys[0], obj, default)
    312     else:
    313         return _get_value_for_keys(

/home/user/uWPS/venv/lib/python3.5/site-packages/marshmallow/utils.py in _get_value_for_key(key, obj, default)
    321         try:
    322             attr = getattr(obj, key)
--> 323             return attr() if callable(attr) else attr
    324         except AttributeError:
    325             return default

TypeError: Required argument 'year' (pos 1) not found

```

What I'm trying to do is to have a way to {,de}serialize a python type. There must be a better way, right?

_Edit_: When I tried a less problematic type, like `int`, the serialized data had the default value for that type, instead of the string representation of the type .. even though I tried adding a `__repr__` to the `Foo` class.

**Comment by carlos-alberto** [comment]
I'm not entirely certain what you're trying to do, but if you define your `Foo` class as follows:

```
class Foo(object):
    def __init__(self, datatype):
        self._datatype = datatype

    @property
    def datatype(self):
        return str(self._datatype)
```

Then calling your dump method gives

```
In [19]: data, errors = schema.dump(Foo(datetime.datetime))

In [20]: data
Out[20]: {'datatype': "<class 'datetime.datetime'>"}
```

**Comment by deckar01** [comment]
@ironhouzi I am able to reproduce the behavior you are describing (thanks for the minimal example). The fact that the exception is escaping the error handler indicates to me that this is a bug.

This is a duplicate of #430 and #397.

**Comment by sloria** [comment]
We intend to fix this in 3.0 (since it requires a backwards-incompatible change). You could try the workarounds suggested in #397

**Comment by sloria** [comment]
Closing as a dupe of #430 ; any further discussion should happen on that issue.
